File: backoffice/src/product/application/ProductRenamer.py
"""
 *
 * Libraries 
 *
"""

from src.product.domain    import ProductName
from src.product.domain    import ProductId
from src.product.domain    import ProductRepository
from src.product.domain    import Product
from src.restaurant.domain import RestaurantId
from src.shared.domain     import ServerInternalErrorException
from src.product.domain    import ProductNameAlreadyCreatedException
from src.restaurant.domain import Restaurant
from src.shared.domain     import DomainEventsPublisher
from src.product.domain    import ProductNotFoundException

"""
 *
 * Class 
 *
"""

class ProductRenamer:

    """
     *
     * Parameters 
     *
    """
 
    __repository     : ProductRepository
    __eventPublisher : DomainEventsPublisher

    """
     *
     * Methods 
     *
    """

    def __init__( 
        self, 
        repository     : ProductRepository,
        eventPublisher : DomainEventsPublisher
    ) -> None:
        self.__repository     = repository
        self.__eventPublisher = eventPublisher

    def rename( 
        self, 
        id           : ProductId, 
        name         : ProductName,
        restaurantId : RestaurantId
    ) -> None:
        # Variables
        product    : Product
        restaurant : Restaurant
        # Code
        product = self.__repository.findByNameAndRestaurant( name, restaurantId )
        if product is not None:
            raise ProductNameAlreadyCreatedException( name )
        product = self.__repository.findByIdAndRestaurant( id, restaurantId )
        if product is None:
            raise ProductNotFoundException( id )
        product.rename( name )
        if not self.__repository.update( product ):
            raise ServerInternalErrorException()
        self.__eventPublisher.publish( product.pullEvents() )

File: preparation/src/product/application/ProductRenamer.cs
/* 
 * 
 * Libraries
 *
*/

using preparation.src.product.domain;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.product.application {

    public class ProductRenamer : DomainEventSubscriber<ProductRenamed> {

        /* 
         * 
         * Attributes
         *
        */

        private ProductRepository repository;

        /* 
         * 
         * Methods
         *
        */

        public ProductRenamer( ProductRepository repository ) {
            this.repository = repository;
        }

        public async Task handleEvent( ProductRenamed domainEvent ) {
            // Variables
            Product? product;
            // Code
            await Task.Run( () => {
                product = repository.findById( domainEvent.getId() );
                if( product == null ) {
                    throw new ProductNotFoundException( domainEvent.getId() );
                }
                product.rename( domainEvent.getName() );
                if( !repository.update( product ) ) {
                    throw new ServerInternalErrorException();
                }
            } );
        }

    }

}